src/main/java/hudson/remoting/RemoteClassLoader.java

```diff
@@ -168,48 +168,80 @@ along with the reference to the initiating ClassLoader (if the initiating ClassL
+
+                                            ClassReference toRef(ClassFile2 cf) {
+                                                int n = cf.classLoader;
+
+                                                ClassLoader cl = classLoaders.get(n);
+                                                if (cl==null)
+                                                     classLoaders.put(n,cl = channel.importedClassLoaders.get(n));
+
+                                                return new ClassReference(cl,cf.image);
+                                            }
+                                        }
+                                        ClassReferenceBuilder crf = new ClassReferenceBuilder();
+
+                                        for (Map.Entry<String,ClassFile2> entry : all.entrySet()) {
+                                            String cn = entry.getKey();
+                                            ClassFile2 cf = entry.getValue();
+                                            ClassReference ref = crf.toRef(cf);
+
+                                            if (cn.equals(name)) {
+                                                cr = ref;
+                                            } else {
+                                                // where we remember the prefetch is sensitive to who references it,
+                                                // because classes need not be transitively visible in Java
+                                                if (cf.referer!=null)
+                                                    ref.rememberIn(cn, crf.toRef(cf.referer).classLoader);
+                                                else
+                                                    ref.rememberIn(cn, this);
+
+                                                LOGGER.log(Level.FINER, "prefetch {0} -> {1}", new Object[]{name, cn});
+                                            }
+
+                                            ref.rememberIn(cn, ref.classLoader);
+                                        }
+                                    }
+                                    break;
+                                } catch (RemotingSystemException x) {
+                                    if (x.getCause() instanceof InterruptedException) {
+                                        // pretend as if this operation is not interruptible.
+                                        // but we need to remember to set the interrupt flag back on
+                                        // before we leave this call.
+                                        interrupted = true;
+                                        continue;   // JENKINS-19453: retry
+                                    }
                                 }
 
-                                ref.rememberIn(cn, ref.classLoader);
+                                // no code is allowed to reach here
```

Comment:
You have pointed me to the fact  I do not propagate exception correctly in the catch. Though, after that is fixed, javac do not let me place the statement here as it will be unreachable statement.